### ISSUE TITLE
How to use createIntl and createIntlCache?

### ISSUE THREAD

**Issue by bensampaio** [formatjs/react-intl#1441]
I have some questions on how `createIntl` and `createIntlCache` work:

* First of all, what is being cached by `createIntlCache`?
* If the language changes dynamically, should I use the same cache or a new cache?

In other words:

```js
const cache = createIntlCache();

const intl = createIntl({
    locale: 'en',
    messages: {},
}, cache);

// At some later point in time
const intl = createIntl({
    locale: 'fr',
    messages: {},
}, cache);
```

or

```js
const intl = createIntl({
    locale: 'en',
    messages: {},
}, createIntlCache());

// At some later point in time
const intl = createIntl({
    locale: 'fr',
    messages: {},
}, createIntlCache());
```

* If the language changes dynamically, should I simply change the `locale` and `messages` properties from the object returned by `createIntl`? Or should I create a new `intl` object with the new `locale` and `messages`?

In other words:

```js
const intl = createIntl({
    locale: 'en',
    messages: {},
});

// At some later point in time
intl.locale = 'fr';
intl.messages = {};
```

or

```js
const intl = createIntl({
    locale: 'en',
    messages: {},
});

// At some later point in time
const intl = createIntl({
    locale: 'fr',
    messages: {},
});
```

* Is it possible / recommended to extend the intl object? For example:

```js
const intl = createIntl({
    locale: 'en',
    messages: {},
    formatWord({ id }) {
        return this.messages[id].replace(/word/g, 'better word');
    },
});
```

Maybe the documentation can be improved so that this becomes a bit more clear.

**Comment by longlho** [comment]
The cache is being used to store `Intl.*` instances and u can re-use the same cache across languages.

1 of the strong use case for caching is SSR where each request can ask for a different language. So the usage would be like:

```tsx
const globalIntlCache = createIntlCache()

const enIntl = createIntl({ locale: 'en' }, globalIntlCache)
const frIntl = createIntl({ locale: 'fr' }, globalIntlCache)
```

**Comment by longlho** [comment]
Doc improvement PRs are always welcome!

**Comment by bensampaio** [comment]
Thanks @longlho, this is exactly what I wanted to know! I would like to contribute with a PR but atm I cannot invest time on that. Maybe this issue will help others in the meantime or maybe I'll come back to this when I have more time.

**Comment by lassegit** [comment]
Could this pattern be used in web app instead of having to import and call `const intl = useIntl()`? Essentially functioning like  https://github.com/alibaba/react-intl-universal?

**Comment by longlho** [comment]
yeah `useIntl` is only for React. If you don't use React you can just use `@formatjs/intl` instead of `react-intl`

**Comment by joshunger** [comment]
@longlho with RSC we're now calling `createIntl` multiple times. Is `createIntl` expensive and should we instead create an array of `IntlShape` for all of the languages we support?

**Comment by longlho** [comment]
You should use the intlCache then. It's not expensive but it'll re-compile messages into AST without the cache.

**Comment by Markyiptw** [comment]
Sorry but I am still confused after reading this issue and #2884. Am I right that messages, or the `intl` instance, are NOT cached? Meaning the following code WON'T work:

```typescript
const cache = createIntlCache();

const intl = createIntl(
  {
    locale: "en",
    messages: { foo: "bar" },
  },
  cache
);

const cached = createIntl(
  {
    locale: "en",
  },
  cache
);

cached.formatMessage({
  id: "foo",
});
```

I am aware that nowhere in the doc suggested this usecase, just that by judging with the name 'cache', and the lack of an api retrieving an instance from a global store (like hooks and provider), one may expect we can call `createIntl` as many time as we want and instances are cached by the `locale`.

**Comment by evert** [comment]
Very helpful ticket! Would be useful in the docs as well!
